### PR TITLE
Template refresh logic for UDP transport

### PR DESCRIPTION
**Pull request by srikartati**

### CODE REVIEW

File: pkg/exporter/process.go [hunk 1]
```diff
@@ -42,35 +42,71 @@ type ExportingProcess struct {
...
 	seqNumber       uint32
+	templateID      uint16
 	set             *entities.Set
 	msg             *entities.MsgBuffer
 	templatesMap    map[uint16]templateValue
+	templateRefCh   chan struct{}
 }
 
-func InitExportingProcess(collectorAddr net.Addr, obsID uint32) (*ExportingProcess, error) {
+// InitExportingProcess takes in collector address(net.Addr format), obsID(observation ID) and tempRefTimeout
```
Comment (zyiou) [inline pkg/exporter/process.go:52]: Maybe we can specify the tempRefTimeout unit (seconds) here

File: pkg/exporter/process.go [hunk 2]
```diff
@@ -42,35 +42,71 @@ type ExportingProcess struct {
...
+			defer ticker.Stop()
+			for {
+				select {
+				case <-expProc.templateRefCh:
+					break
+				case <-ticker.C:
+					err := expProc.sendRefreshedTemplates()
+					if err != nil {
+						// Other option is sending messages through channel to library consumers
+						log.Printf("Error when sending refreshed templates. Closing the connection to IPFIX controller")
```
Comment (zyiou) [inline pkg/exporter/process.go:92]: Will it be better if we also print the error message in the log?

File: pkg/exporter/process_test.go [hunk 3]
```diff
@@ -187,21 +207,30 @@ func TestExportingProcess_SendingDataRecordToLocalTCPServer(t *testing.T) {
...
-
-	// [Only for testing] Ensure corresponding template exists in the exporting process before sending data
-	templateID := exporter.AddTemplate()
-	// Hardcoding 8-bytes min data record length for testing purposes instead of creating template record
-	exporter.updateTemplate(templateID, &[]string{"sourceIPv4Address", "destinationIPv4Address"}, 8)
 	// Add data to exporting process
 	reg := registry.NewIanaRegistry()
 	reg.LoadRegistry()
 
+	// [Only for testing] Ensure corresponding template exists in the exporting process before sending data
```
Comment (zyiou) [inline pkg/exporter/process_test.go:220]: This line of comment can be deleted (I added it for the same purpose of line 231)

File: pkg/exporter/process_test.go [hunk 4]
```diff
@@ -256,19 +285,29 @@ func TestExportingProcess_SendingDataRecordToLocalUDPServer(t *testing.T) {
...
 	t.Logf("Created exporter connecting to local server with address: %s", conn.LocalAddr().String())
 
-	// [Only for testing] Ensure corresponding template exists in the exporting process before sending data
-	templateID := exporter.AddTemplate()
-	// Hardcoding 8-bytes min data record length for testing purposes instead of creating template record
-	exporter.updateTemplate(templateID, &[]string{"sourceIPv4Address", "destinationIPv4Address"}, 8)
 	// Add data to exporting process
 	reg := registry.NewIanaRegistry()
 	reg.LoadRegistry()
+	// [Only for testing] Ensure corresponding template exists in the exporting process before sending data
```
Comment (zyiou) [inline pkg/exporter/process_test.go:297]: Same as above one. This line can be deleted. Thanks!

**Review by zyiou — APPROVED** [review]

File: pkg/exporter/process.go (hunk 2 above)
Comment (srikartati) [reply pkg/exporter/process.go:92]: There is no error message in log library. Only options are fatal, panic and print helpers.

File: pkg/exporter/process_test.go (hunk 3 above)
Comment (srikartati) [reply pkg/exporter/process_test.go:220]: Actually we still need as we are creating a template and updating template map. This is not required if we send a template record.

File: pkg/exporter/process.go (hunk 1 above)
Comment (srikartati) [reply pkg/exporter/process.go:52]: done

File: pkg/exporter/process_test.go (hunk 3 above)
Comment (zyiou) [reply pkg/exporter/process_test.go:220]: Got it. Thanks

File: pkg/exporter/process.go (hunk 2 above)
Comment (zyiou) [reply pkg/exporter/process.go:92]: I mean something like `log.Printf("Error xxx %v", err)` so that we can have a better idea of what the err is.

**Comment by antoninbas** [comment]
@srikartati @zyiou just a comment on logging. It would be good to make sure the library integrates nicely with Antrea, so maybe you could look into using klog right away. An alternative is to use `sirupsen/logrus` and we can do the same trick we used for libOpenflow (https://github.com/vmware-tanzu/antrea/blob/master/pkg/ovs/openflow/logs.go), but really I don't see a reason not to go with klog.

**Comment by srikartati** [comment]
> @srikartati @zyiou just a comment on logging. It would be good to make sure the library integrates nicely with Antrea, so maybe you could look into using klog right away. An alternative is to use `sirupsen/logrus` and we can do the same trick we used for libOpenflow (https://github.com/vmware-tanzu/antrea/blob/master/pkg/ovs/openflow/logs.go), but really I don't see a reason not to go with klog.

Yes going towards klog is good idea.
Created an issue in vmware/go-ipfix https://github.com/vmware/go-ipfix/issues/4

**Comment by srikartati** [comment]
Merging this one for now.

**Pull request merged**